### PR TITLE
feat / Eterbase exchange connector

### PR DESCRIPTION
**Pull request by eterbase**
**Before submitting this PR, please make sure**:
- [x] Your code builds clean without any errors or warnings
- [x] You are using approved title ("feat/", "fix/", "docs/", "refactor/", etc)

**Optional**:
- Related Github issue:
- Related Clubhouse Story:

**A description of the changes proposed in the pull request**:

### CODE REVIEW

**Comment by mifeng** [comment]
It looks like the user is asked to specify their fee tier when they connect to Eterbase - should this be an input?

Also, we are currently refactoring and simplifying how Hummingbot handles fees. We'll let you know when the new version is done, in about a week.

**Comment by dennisocana** [comment]
Unit tests for `test_eterbase_market.py` passed.

```
test_cancel_all (test.integration.test_eterbase_market.EterbaseMarketUnitTest) ... ok
test_cancel_order (test.integration.test_eterbase_market.EterbaseMarketUnitTest) ... ok
test_get_fee (test.integration.test_eterbase_market.EterbaseMarketUnitTest) ... ok
test_limit_buy (test.integration.test_eterbase_market.EterbaseMarketUnitTest) ... ok
test_limit_sell (test.integration.test_eterbase_market.EterbaseMarketUnitTest) ... ok
test_list_orders (test.integration.test_eterbase_market.EterbaseMarketUnitTest) ... SKIP: List order test requires manual action.
test_market_buy (test.integration.test_eterbase_market.EterbaseMarketUnitTest) ... ok
test_market_sell (test.integration.test_eterbase_market.EterbaseMarketUnitTest) ... ok
test_order_fill_record (test.integration.test_eterbase_market.EterbaseMarketUnitTest) ... ok
test_orders_saving_and_restoration (test.integration.test_eterbase_market.EterbaseMarketUnitTest) ... ok
test_withdraw (test.integration.test_eterbase_market.EterbaseMarketUnitTest) ... SKIP: Withdraw test requires manual action.

----------------------------------------------------------------------
Ran 11 tests in 15.015s

OK (SKIP=2)
Initializing Eterbase market... this will take about a minute.
Ready.
close out the current market and start another market
```

**Comment by dennisocana** [comment]
## Delisted Pairs

Trading pairs with `"state": "Delisted"` are showing in trading pair auto-complete and will result to errors when used as `market`.

putty_P9LTnIGaBc

putty_jOxaHdkf3l

```
2020-06-08 11:44:19,496 - 31383 - hummingbot.market.eterbase.eterbase_api_order_book_data_source - NETWORK - Unexpected error with WebSocket connection.
Traceback (most recent call last):
  File "/home/ubuntu/hummingbot/hummingbot/market/eterbase/eterbase_api_order_book_data_source.py", line 342, in listen_for_order_book_snapshots
    snapshot: Dict[str, any] = await self.get_snapshot(client, trading_pair)
  File "/home/ubuntu/hummingbot/hummingbot/market/eterbase/eterbase_api_order_book_data_source.py", line 192, in get_snapshot
    raise IOError(f"Error fetching Eterbase market snapshot for marketId: {market_id}. "
OSError: Error fetching Eterbase market snapshot for marketId: 40. HTTP status is 400.
```

```
2020-06-08 11:44:20,141 - 31383 - hummingbot.market.eterbase.eterbase_api_order_book_data_source - NETWORK - Unexpected error with WebSocket connection.
Traceback (most recent call last):
  File "/home/ubuntu/hummingbot/hummingbot/market/eterbase/eterbase_api_order_book_data_source.py", line 290, in listen_for_order_book_diffs
    marketIds.append(tp_map_mrktid[tp])
KeyError: 'USDTUSDC'
```

```
"id": 40,
"symbol": "USDTUSDC",
"base": "USDT",
"quote": "USDC",
"priceSigDigs": 5,
"qtySigDigs": 8,
"costSigDigs": 8,
"verificationLevelUser": 0,
"verificationLevelCorporate": 10,
"group": "SC",
"tradingRules": [
    {
        "attribute": "Qty",
        "condition": "Min",
        "value": 1
    },
    {
        "attribute": "Qty",
        "condition": "Max",
        "value": 210000
    },
    {
        "attribute": "Cost",
        "condition": "Min",
        "value": 1
    },
    {
        "attribute": "Cost",
        "condition": "Max",
        "value": 210000
    }
],
"allowedOrderTypes": [
    1,
    2,
    3,
    4
],
"state": "Delisted"
```

**Comment by dennisocana** [comment]
@eterbase fix confirmed on all bugs mentioned.

**Comment by dennisocana** [comment]
@eterbase in pure market making, when `inventory_skew_enabled` is set to `False` the actual order in the exchange does not match the order amount in Hummingbot. Please see screenshots below.

putty_PCg0VENxDy
AeTfwNPPmk

Does not happen when inventory skew is enabled.

putty_RrfdZIiC8B

YYgRR7PeiA

File: hummingbot/market/eterbase/eterbase_order_book_tracker.py [hunk 1]
```diff
@@ -0,0 +1,229 @@
...
+        new_trading_pairs: Set[str] = available_trading_pairs - tracking_trading_pairs
+        deleted_trading_pairs: Set[str] = tracking_trading_pairs - available_trading_pairs
+
+        for trading_pair in new_trading_pairs:
+            order_book_tracker_entry: EterbaseOrderBookTrackerEntry = available_pairs[trading_pair]
+            self._active_order_trackers[trading_pair] = order_book_tracker_entry.active_order_tracker
+            self._order_books[trading_pair] = order_book_tracker_entry.order_book
+            self._tracking_message_queues[trading_pair] = asyncio.Queue()
+            self._tracking_tasks[trading_pair] = safe_ensure_future(self._track_single_book(trading_pair))
+            self.logger().info("Started order book tracking for %s.", trading_pair)
```
Comment (dennisocana) [inline hummingbot/market/eterbase/eterbase_order_book_tracker.py:100]: Minor typo in order book tracking log message.

putty_RFPmlwxh2I

**Comment by dennisocana** [comment]
@eterbase pure market making parameter `order_optimization_enabled` is not functioning as expected. Screenshot below shows the order when this feature is disabled.

PwLP4Q2Fpi putty_8p7eyu3XRe 

When enabled, the expected behavior is it puts the order just one tick above the best order. Expected sell order price should be 243.42

BJMRdULJM7 putty_ucXUSyqicu

**Comment by dennisocana** [comment]
## Intermittent connection issues

@eterbase ran a bot overnight and ran into intermittent connection issues that would sometimes cause not to cancel active orders.

Log file : [logs_testing_eterbase.log](https://github.com/CoinAlpha/hummingbot/files/4763907/logs_testing_eterbase.log)


putty_e6ykhgjKOB
Code_uERkvY2LP6
Code_rcCObKbT8a
Code_dbGy2BFaTN

File: hummingbot/market/eterbase/eterbase_active_order_tracker.pyx [hunk 2]
```diff
@@ -0,0 +1,328 @@
...
+
+import logging
+import numpy as np
+from decimal import Decimal
+from typing import Dict
+
+from hummingbot.logger import HummingbotLogger
+from hummingbot.core.data_type.order_book_row import OrderBookRow
+
+_cbpaot_logger = None
```
Comment (kmcgowan2000) [inline hummingbot/market/eterbase/eterbase_active_order_tracker.pyx:12]: `_cbpaot_logger` is the coinbase pro active order tracker logger – yours could be `_eaot_logger`

Comment (kmcgowan2000) [reply hummingbot/market/eterbase/eterbase_active_order_tracker.pyx:12]: (same goes for all loggers)

**Comment by dennisocana** [comment]
Confirmed fix on the issues previously mentioned :
- Order amount in Eterbase does not match in client - resolved
- Typo in order book tracking log message - resolved
- Order optimization in Pure-MM - resolved

See comments below on other bugs found while testing overnight.

**Comment by dennisocana** [comment]
## Creating orders on low balance

The bot keeps trying to create an order size based on the remaining available balance ignoring the specified value for `order_amount`. To replicate this bug, turn `inventory_skew_enabled` to `False` and let one side to have insufficient assets to place an order.

For example, trade on ETH-USDT pair with `order_amount` set to `0.5` and have only 2 USDT. The bot will attempt to place a buy order of 0.009 ETH.

Log file : [eterbase_2.log](https://github.com/CoinAlpha/hummingbot/files/4796413/eterbase_2.log)

K6pW6UF7c5

**Comment by dennisocana** [comment]
## Order stuck in client

Might be related to the connection issue. An order got stuck in Hummingbot client so it keeps flooding with "Cancelling the limit order" message and no new orders are created. It's in the same log file as the comment above.

putty_r5yKCkWeZm
Code_dPsRyvkd8i

**Comment by dennisocana** [comment]
## Error fetching status update for orders

putty_WlP8YKsxV4

Log file : [eterbase_error_fetching_status_updates.log](https://github.com/CoinAlpha/hummingbot/files/4829441/eterbase_error_fetching_status_updates.log)

**Comment by dennisocana** [comment]
OK to test

**Comment by dennisocana** [comment]
test this please

**Comment by carlolm** [comment]
OK to test

**Comment by dennisocana** [comment]
@eterbase please see Jenkins unit test results below.

```
+ /root/miniconda/envs/hummingbot/bin/nosetests -d -v test/test_arbitrage.py test/test_celo_arb.py test/test_config_security.py test/test_config_templates.py test/test_cross_exchange_market_making.py test/test_inventory_skew_calculator.py test/test_perform_trade.py test/test_pmm.py test/test_pmm_ping_pong.py test/test_pmm_refresh_tolerance.py test/test_pmm_take_if_cross.py test/test_simple_trade.py test/test_trade_performance_analysis.py test/test_twap.py test/test_vwap.py
fatal: ref HEAD is not a symbolic ref
test_arbitrage_not_profitable (test.test_arbitrage.ArbitrageUnitTest) ... ok
test_arbitrage_profitable (test.test_arbitrage.ArbitrageUnitTest) ... ok
test_asset_limit (test.test_arbitrage.ArbitrageUnitTest) ... ok
test_find_best_profitable_amount (test.test_arbitrage.ArbitrageUnitTest) ... ok
test_find_profitable_arbitrage_orders (test.test_arbitrage.ArbitrageUnitTest) ... ok
test_min_profitability_limit_1 (test.test_arbitrage.ArbitrageUnitTest) ... ok
test_min_profitability_limit_2 (test.test_arbitrage.ArbitrageUnitTest) ... ok
test_ready_for_new_orders (test.test_arbitrage.ArbitrageUnitTest) ... ok
Order book bids ... ok
test_no_profitable_trade (test.test_celo_arb.CeloArbUnitTest) ... ok
test_profitable_but_insufficient_balance (test.test_celo_arb.CeloArbUnitTest) ... ok
test_profitable_celo_buy_trade (test.test_celo_arb.CeloArbUnitTest) ... ok
test_profitable_celo_sell_trade (test.test_celo_arb.CeloArbUnitTest) ... ok
test_existing_password (test.test_config_security.ConfigSecurityExistingPasswordUnitTest) ... ok
test_new_password_process (test.test_config_security.ConfigSecurityNewPasswordUnitTest) ... ok
test_global_config_prompt_exists (test.test_config_templates.ConfigTemplatesUnitTest) ... ok
test_global_config_template_complete (test.test_config_templates.ConfigTemplatesUnitTest) ... FAIL
test_strategy_config_template_complete (test.test_config_templates.ConfigTemplatesUnitTest) ... ok
test_both_sides_profitable (test.test_cross_exchange_market_making.HedgedMarketMakingUnitTest) ... ok
test_empty_maker_orderbook (test.test_cross_exchange_market_making.HedgedMarketMakingUnitTest) ... ok
test_maker_price (test.test_cross_exchange_market_making.HedgedMarketMakingUnitTest) ... ok
test_market_became_narrower (test.test_cross_exchange_market_making.HedgedMarketMakingUnitTest) ... ok
test_market_became_wider (test.test_cross_exchange_market_making.HedgedMarketMakingUnitTest) ... ok
test_order_fills_after_cancellation (test.test_cross_exchange_market_making.HedgedMarketMakingUnitTest) ... ok
test_price_and_size_limit_calculation (test.test_cross_exchange_market_making.HedgedMarketMakingUnitTest) ... ok
test_top_depth_tolerance (test.test_cross_exchange_market_making.HedgedMarketMakingUnitTest) ... ok
test_with_adjust_orders_disabled (test.test_cross_exchange_market_making.HedgedMarketMakingUnitTest) ... ok
test_with_adjust_orders_enabled (test.test_cross_exchange_market_making.HedgedMarketMakingUnitTest) ... ok
test_with_conversion (test.test_cross_exchange_market_making.HedgedMarketMakingUnitTest) ... ok
test_balanced_portfolio (test.test_inventory_skew_calculator.InventorySkewCalculatorUnitTest) ... ok
test_cap_on_max_base_range (test.test_inventory_skew_calculator.InventorySkewCalculatorUnitTest) ... ok
test_empty_portfolio (test.test_inventory_skew_calculator.InventorySkewCalculatorUnitTest) ... ok
test_heavily_skewed_portfolio (test.test_inventory_skew_calculator.InventorySkewCalculatorUnitTest) ... ok
test_moderately_skewed_portfolio (test.test_inventory_skew_calculator.InventorySkewCalculatorUnitTest) ... ok
test_limit_buy_order (test.test_perform_trade.PerformTradeUnitTest) ... ok
test_limit_sell_order (test.test_perform_trade.PerformTradeUnitTest) ... ok
test_market_buy_order (test.test_perform_trade.PerformTradeUnitTest) ... ok
test_market_sell_order (test.test_perform_trade.PerformTradeUnitTest) ... ok
test_order_filled_events (test.test_perform_trade.PerformTradeUnitTest) ... ok
test_with_insufficient_balance (test.test_perform_trade.PerformTradeUnitTest) ... ok
test_add_transaction_costs (test.test_pmm.PMMUnitTest) ... ok
test_apply_budget_constraint_to_proposal (test.test_pmm.PMMUnitTest) ... ok
test_basic_multiple_levels (test.test_pmm.PMMUnitTest) ... ok
test_basic_one_level (test.test_pmm.PMMUnitTest) ... ok
test_config_spread_on_the_fly_multiple_orders (test.test_pmm.PMMUnitTest) ... ok
test_external_exchange_price_source (test.test_pmm.PMMUnitTest) ... ok
test_external_exchange_price_source_empty_orderbook (test.test_pmm.PMMUnitTest) ... ok
test_filled_order_delay (test.test_pmm.PMMUnitTest) ... ok
test_filled_order_delay_mulitiple_orders (test.test_pmm.PMMUnitTest) ... ok
test_hanging_orders (test.test_pmm.PMMUnitTest) ... ok
test_hanging_orders_multiple_orders (test.test_pmm.PMMUnitTest) ... ok
test_inventory_skew (test.test_pmm.PMMUnitTest) ... ok
test_inventory_skew_multiple_orders (test.test_pmm.PMMUnitTest) ... ok
test_inventory_skew_multiple_orders_status (test.test_pmm.PMMUnitTest) ... ok
test_market_became_narrower (test.test_pmm.PMMUnitTest) ... ok
test_market_become_wider (test.test_pmm.PMMUnitTest) ... ok
test_multi_order_external_exchange_price_source (test.test_pmm.PMMUnitTest) ... ok
test_multi_order_external_exchange_price_source_empty_order_book (test.test_pmm.PMMUnitTest) ... ok
test_order_optimization (test.test_pmm.PMMUnitTest) ... ok
test_price_band_price_ceiling_breach (test.test_pmm.PMMUnitTest) ... ok
test_price_band_price_floor_breach (test.test_pmm.PMMUnitTest) ... ok
test_minimum_spread_param (test.test_pmm.PureMarketMakingMinimumSpreadUnitTest) ... ok
test_multiple_orders_ping_pong (test.test_pmm_ping_pong.PMMRefreshToleranceUnitTest) ... ok
test_strategy_ping_pong_on_ask_fill (test.test_pmm_ping_pong.PMMRefreshToleranceUnitTest) ... ok
test_strategy_ping_pong_on_bid_fill (test.test_pmm_ping_pong.PMMRefreshToleranceUnitTest) ... ok
test_active_orders_are_cancelled_when_mid_price_moves (test.test_pmm_refresh_tolerance.PMMRefreshToleranceUnitTest) ... ok
test_active_orders_are_kept_when_within_tolerance (test.test_pmm_refresh_tolerance.PMMRefreshToleranceUnitTest) ... ok
test_hanging_orders_multiple_orders_with_refresh_tolerance (test.test_pmm_refresh_tolerance.PMMRefreshToleranceUnitTest) ... ok
test_multi_levels_active_orders_are_cancelled_when_mid_price_moves (test.test_pmm_refresh_tolerance.PMMRefreshToleranceUnitTest) ... ok
test_multiple_active_orders_are_kept_when_within_tolerance (test.test_pmm_refresh_tolerance.PMMRefreshToleranceUnitTest) ... ok
test_strategy_take_if_crossed_ask_order (test.test_pmm_take_if_cross.PureMMTakeIfCrossUnitTest) ... ok
test_strategy_take_if_crossed_bid_order (test.test_pmm_take_if_cross.PureMMTakeIfCrossUnitTest) ... ok
test_limit_buy_order (test.test_simple_trade.SimpleTradeUnitTest) ... ok
test_limit_sell_order (test.test_simple_trade.SimpleTradeUnitTest) ... ok
test_market_buy_order (test.test_simple_trade.SimpleTradeUnitTest) ... ok
test_market_sell_order (test.test_simple_trade.SimpleTradeUnitTest) ... ok
test_order_filled_events (test.test_simple_trade.SimpleTradeUnitTest) ... ok
test_with_insufficient_balance (test.test_simple_trade.SimpleTradeUnitTest) ... ok
test_calculate_asset_delta_from_trades (test.test_trade_performance_analysis.TestTradePerformanceAnalysis) ... ok
test_calculate_trade_performance (test.test_trade_performance_analysis.TestTradePerformanceAnalysis) ... ok
test_calculate_trade_quote_delta_with_fees (test.test_trade_performance_analysis.TestTradePerformanceAnalysis) ... ok
test_multiple_market (test.test_trade_performance_analysis.TestTradePerformanceAnalysis) ... ok
test_limit_buy_order (test.test_twap.TWAPUnitTest) ... ok
test_limit_sell_order (test.test_twap.TWAPUnitTest) ... ok
test_market_buy_order (test.test_twap.TWAPUnitTest) ... ok
test_market_sell_order (test.test_twap.TWAPUnitTest) ... ok
test_order_filled_events (test.test_twap.TWAPUnitTest) ... ok
test_with_insufficient_balance (test.test_twap.TWAPUnitTest) ... ok
test_limit_buy_order (test.test_vwap.TWAPUnitTest) ... ok
test_limit_sell_order (test.test_vwap.TWAPUnitTest) ... ok
test_market_buy_order (test.test_vwap.TWAPUnitTest) ... ok
test_market_sell_order (test.test_vwap.TWAPUnitTest) ... ok
test_order_filled_events (test.test_vwap.TWAPUnitTest) ... ok
test_with_insufficient_balance (test.test_vwap.TWAPUnitTest) ... ok

======================================================================
FAIL: test_global_config_template_complete (test.test_config_templates.ConfigTemplatesUnitTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/var/jenkins_home/workspace/hb-test_pull-request/test/test_config_templates.py", line 39, in test_global_config_template_complete
    self.assertTrue(key in template_data, f"{key} not in {global_config_template_path}")
AssertionError: False is not true : eterbase_api_key not in /var/jenkins_home/workspace/hb-test_pull-request/hummingbot/templates/conf_global_TEMPLATE.yml
    'False is not true : eterbase_api_key not in /var/jenkins_home/workspace/hb-test_pull-request/hummingbot/templates/conf_global_TEMPLATE.yml' = self._formatMessage('False is not true : eterbase_api_key not in /var/jenkins_home/workspace/hb-test_pull-request/hummingbot/templates/conf_global_TEMPLATE.yml', "%s is not true" % safe_repr(False))
>>  raise self.failureException('False is not true : eterbase_api_key not in /var/jenkins_home/workspace/hb-test_pull-request/hummingbot/templates/conf_global_TEMPLATE.yml')
    

----------------------------------------------------------------------
Ran 94 tests in 5.198s

FAILED (failures=1)
```

**Pull request merged**